apps/src/netsim/NetSimNodeClient.js

```diff
@@ -0,0 +1,257 @@
+ * either by our own client or somebody else's.  Is a NetSimEntity, meaning
+ * it wraps a row in the node table and provides functionality around it.
+ */
+
+/* jshint
+ funcscope: true,
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+
+ maxlen: 90,
+ maxparams: 3,
+ maxstatements: 200
+ */
+'use strict';
+
+var superClass = require('./NetSimNode');
+var NetSimEntity = require('./NetSimEntity');
+
+/**
+ * @param {!netsimInstance} instance
+ * @param {Object} [clientRow] - Lobby row for this router.
+ * @constructor
+ * @augments NetSimNode
+ */
+var NetSimNodeClient = function (instance, clientRow) {
+  superClass.call(this, instance, clientRow);
+
+  /**
+   * How long (in milliseconds) this entity is allowed to remain in
+   * storage without being cleaned up.
+   * @type {number}
+   * @override
+   */
+  this.ENTITY_TIMEOUT_MS = 30000;
+
+  /**
+   * How often (in milliseconds) this entity's status should be pushed
+   * to the server to keep the row active.
+   * @type {number}
+   * @override
+   */
+  this.ENTITY_KEEPALIVE_MS = 2000;
+
+  /**
+   * Client nodes can only have one wire at a time.
+   * @type {NetSimWire}
+   */
+  this.myWire = null;
+
+  /**
+   * Client nodes can be connected to a router, which they will
+   * help to simulate.
+   * @type {NetSimRouter}
+   */
+  this.myRouter = null;
+};
+NetSimNodeClient.prototype = Object.create(superClass.prototype);
+NetSimNodeClient.prototype.constructor = NetSimNodeClient;
+module.exports = NetSimNodeClient;
+
+/**
+ * Static async creation method. See NetSimEntity.create().
+ * @param {!netsimInstance} instance
+ * @param {function} [onComplete] - Method that will be given the
+ *        created entity, or null if entity creation failed.
+ */
+NetSimNodeClient.create = function (instance, onComplete) {
+  NetSimEntity.create(NetSimNodeClient, instance, onComplete);
+};
+
+/**
+ * @inheritdoc
+ */
+NetSimNodeClient.prototype.getNodeType = function () {
+  return NetSimNodeClient.getNodeType();
+};
+NetSimNodeClient.getNodeType = function () {
+  return 'user';
+};
+
+/**
+ * @inheritdoc
+ */
+NetSimNodeClient.prototype.getStatus = function () {
+  return this.status_ ? this.status_ : 'Online';
+};
+
+NetSimNodeClient.prototype.setDisplayName = function (displayName) {
+  this.displayName_ = displayName;
+};
+
+/**
+ * Our client tick can also tick its connected wire and its remote client.
+ * @param {!RunLoop.Clock} clock
+ */
+NetSimNodeClient.prototype.tick = function (clock) {
+  superClass.prototype.tick.call(this, clock);
+
+  if (this.myWire) {
+    this.myWire.tick(clock);
+  }
+
+  if (this.myRouter) {
+    this.myRouter.tick(clock);
+  }
+};
+
+/**
+ * If a client update fails, should attempt an automatic reconnect.
+ * @param {function} [onComplete]
+ * @param {boolean} [autoReconnect=true]
+ */
+NetSimNodeClient.prototype.update = function (onComplete, autoReconnect) {
+  if (!onComplete) {
```

Comment:
How about this? onComplete !== undefined && onComplete();

I would hesitate to go with something that jshint doesn't like so if you or
jshint don't like it then feel free to change it back... I was just looking
for ways to make the code a little denser.

On Thu, Feb 5, 2015 at 3:45 PM, Brad Buchanan <notifications@github.com>
wrote:

> In apps/src/netsim/NetSimNodeClient.js
> <https://github.com/code-dot-org/code-dot-org/pull/1207#discussion_r24211374>
> :
>
> > +  if (this.myWire) {
> > +    this.myWire.tick(clock);
> > +  }
> > +
> > +  if (this.myRouter) {
> > +    this.myRouter.tick(clock);
> > +  }
> > +};
> > +
> > +/**
> > + * If a client update fails, should attempt an automatic reconnect.
> > + * @param {function} [onComplete]
> > + * @param {boolean} [autoReconnect=true]
> > + */
> > +NetSimNodeClient.prototype.update = function (onComplete, autoReconnect) {
> > +  if (!onComplete) {
>
> Hmm... I made this change (using onComplete && onComplete()) but JSHint
> complains about it:
>
> Expected an assignment or function call and instead saw an expression.
>
> Should I change it back? Adjust JSHint settings?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/code-dot-org/code-dot-org/pull/1207/files#r24211374>.
>